### PR TITLE
Open source build of Visual Studio Code

### PR DESCRIPTION
**Pull request by amtlib-dot-dll**
Microsoft [does not provide](http://code.visualstudio.com/license) official binary of Visual Studio Code, and [the existing repo](https://github.com/flathub/com.visualstudio.code) in flathub is based on the proprietary version.

### CODE REVIEW

**Comment by nedrichards** [comment]
I'm confused about your point. Microsoft doesn't allow redistribution of VS Code in their licence. Which is why Flathub uses extra-data so the user obtains the binary directly. No redistribution takes place.

I'm far from against building things from source - but in this case the OSS version is a different app with a different app id - I know since Endless used to [build it from source](https://github.com/endlessm/vscode-flatpak) as well! (we helped contribute flatpak building machinery upstream that has since rotted).

**Comment by nedrichards** [comment]
bot, build com.visualstudio.code-oss

File: com.visualstudio.code-oss.appdata.xml [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+	</description>
+	<screenshots>
+		<screenshot type="default">
+			<image>https://code.visualstudio.com/home/home-screenshot-linux-lg.png</image>
+			<caption>Editing TypeScript and searching for extensions</caption>
+		</screenshot>
+	</screenshots>
+	<categories>
+    	<category>Development</category>
+		<category>IDE</category>
```
Comment (nedrichards) [inline com.visualstudio.code-oss.appdata.xml]: Indenting

File: com.visualstudio.code-oss.appdata.xml [hunk 2]
```diff
@@ -0,0 +1,49 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<component type="desktop">
+	<id>com.visualstudio.code-oss.desktop</id>
+	<metadata_license>GFDL-1.3</metadata_license>
```
Comment (nedrichards) [inline com.visualstudio.code-oss.appdata.xml]: Is this using GFDL because it's derived from the upstream app data?

**Comment by nedrichards** [comment]
I'd be a lot happier if the [appata was being pushed upstream](https://github.com/Microsoft/vscode/blob/master/resources/linux/code.appdata.xml) as well, since theirs could do with some improvements!

**Comment by nedrichards** [comment]
Again, for the desktop file - I assume you use your own fork because of the translations? Would be best to get that upstream.

File: com.visualstudio.code-oss.json [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+                {
+                    "type": "git",
+                    "url": "https://github.com/Microsoft/vscode.git",
+                    "tag": "1.20.1",
+                    "commit": "f88bbf9137d24d36d968ea6b2911786bfe103002",
+                    "dest": "vscode"
+                },
+                {
+                    "type": "file",
+                    "url": "https://upload.wikimedia.org/wikipedia/commons/2/2d/Visual_Studio_Code_1.18_icon.svg",
```
Comment (nedrichards) [inline com.visualstudio.code-oss.json]: There's an upstream icon for the OSS version and it's not this. If we're going to have two versions in flathub they can't be confusing.

File: flathub.json [hunk 4]
```diff
@@ -0,0 +1,5 @@
+{
+    "only-arches": [
+        "x86_64"
```
Comment (nedrichards) [inline flathub.json]: Why? Surely the whole point of building from source is to enable other architectures? Electron supports all our architectures.

**Comment by nedrichards** [comment]
The build feedback is currently broken but: https://flathub.org/builds/#/builders/43/builds/3

> ]2;flatpak-builder: Downloading vscodeFailed to download sources: module vscode: Can't find file at SHASUMS256.txt-1.7.0

File: com.visualstudio.code-oss.appdata.xml (hunk 2 above)
Comment (amtlib-dot-dll) [reply com.visualstudio.code-oss.appdata.xml]: No, it is written by me. I did not notice the upstream app data and this may be a better option, thank you!

Comment (amtlib-dot-dll) [reply com.visualstudio.code-oss.appdata.xml]: The translations are copied from https://aur.archlinux.org/cgit/aur.git/tree/?h=code

File: com.visualstudio.code-oss.json (hunk 3 above)
Comment (amtlib-dot-dll) [reply com.visualstudio.code-oss.json]: OK, I will change

File: flathub.json (hunk 4 above)
Comment (amtlib-dot-dll) [reply flathub.json]: I have only tested 64-bit build and I don't know whether my `build.sh` can build under 32-bit due to the native modules

**Comment by amtlib-dot-dll** [comment]
Sorry, I forgot to upload these two files yesterday.

**Comment by amtlib-dot-dll** [comment]
bot, build com.visualstudio.code-oss. Can I trigger the build?

File: flathub.json (hunk 4 above)
Comment (nedrichards) [reply flathub.json]: hmmm. does the yarn generator do per arch dependencies for that sort of thing?

As for everything else - it's pretty easy for us to check on the builders - either with this or a test PR

**Comment by nedrichards** [comment]
You can't trigger builds before the app has been added to flathub I'm afraid - but once the repo is created you can trigger builds in PR for testing.

**Comment by nedrichards** [comment]
bot, build com.visualstudio.code-oss

**Comment by nedrichards** [comment]
Should cleanup all of imagemagick at the end after you use it.

**Comment by amtlib-dot-dll** [comment]
Sorry, the build fails on my machine, and I am figuring out why :joy: :joy:

**Comment by amtlib-dot-dll** [comment]
The build you triggered are still downloading files :joy: and why there are 2w+ lines of output :joy: amazing

**Comment by nedrichards** [comment]
@amtlib-dot-dll it [just finished](https://flathub.org/builds/#/builders/61/builds/4):

```Running: install -Dm644 vscode/resources/linux/code.png /app/share/icons/hicolor/1024x1024/apps/com.visualstudio.code-oss.png
FB: Running: flatpak build --die-with-parent --nofilesystem=host --filesystem=/mnt/builder/worker/build-x86_64-5/com.visualstudio.code-oss/.flatpak-builder/build/vscode-1 --bind-mount=/run/build/vscode=/mnt/builder/worker/build-x86_64-5/com.visualstudio.code-oss/.flatpak-builder/build/vscode-1 --build-dir=/run/build/vscode --env=PATH=/app/bin:/usr/bin --env=LD_LIBRARY_PATH=/app/lib --env=PKG_CONFIG_PATH=/app/lib/pkgconfig:/app/share/pkgconfig:/usr/lib/pkgconfig:/usr/share/pkgconfig /mnt/builder/worker/build-x86_64-5/com.visualstudio.code-oss/.flatpak-builder/rofiles/rofiles-ECMYwD /bin/sh -c 'install -Dm644 vscode/resources/linux/code.png /app/share/icons/hicolor/1024x1024/apps/com.visualstudio.code-oss.png'
Running: for resolution in 512x512 256x256 128x128 64x64; do convert -resize $resolution vscode/resources/linux/code.png $resolution.png; install -Dm644 $resolution.png /app/share/icons/hicolor/$resolution/apps/com.visualstudio.code-oss.png; done
FB: Running: flatpak build --die-with-parent --nofilesystem=host --filesystem=/mnt/builder/worker/build-x86_64-5/com.visualstudio.code-oss/.flatpak-builder/build/vscode-1 --bind-mount=/run/build/vscode=/mnt/builder/worker/build-x86_64-5/com.visualstudio.code-oss/.flatpak-builder/build/vscode-1 --build-dir=/run/build/vscode --env=PATH=/app/bin:/usr/bin --env=LD_LIBRARY_PATH=/app/lib --env=PKG_CONFIG_PATH=/app/lib/pkgconfig:/app/share/pkgconfig:/usr/lib/pkgconfig:/usr/share/pkgconfig /mnt/builder/worker/build-x86_64-5/com.visualstudio.code-oss/.flatpak-builder/rofiles/rofiles-ECMYwD /bin/sh -c 'for resolution in 512x512 256x256 128x128 64x64; do convert -resize $resolution vscode/resources/linux/code.png $resolution.png; install -Dm644 $resolution.png /app/share/icons/hicolor/$resolution/apps/com.visualstudio.code-oss.png; done'
convert: no decode delegate for this image format `PNG' @ error/constitute.c/ReadImage/509.
convert: no images defined `512x512.png' @ error/convert.c/ConvertImageCommand/3275.
install: cannot stat '512x512.png': No such file or directory
convert: no decode delegate for this image format `PNG' @ error/constitute.c/ReadImage/509.
convert: no images defined `256x256.png' @ error/convert.c/ConvertImageCommand/3275.
install: cannot stat '256x256.png': No such file or directory
convert: no decode delegate for this image format `PNG' @ error/constitute.c/ReadImage/509.
convert: no images defined `128x128.png' @ error/convert.c/ConvertImageCommand/3275.
install: cannot stat '128x128.png': No such file or directory
convert: no decode delegate for this image format `PNG' @ error/constitute.c/ReadImage/509.
convert: no images defined `64x64.png' @ error/convert.c/ConvertImageCommand/3275.
install: cannot stat '64x64.png': No such file or directory```

looks like your imagemagick doesn't know what a PNG is ;-)

**Comment by nedrichards** [comment]
bot, build com.visualstudio.code-oss

**Comment by amtlib-dot-dll** [comment]
:thinking: What commit is this build against? Apart from the imagemagick issue my builder fails on building the nodejs modules :joy:

**Comment by amtlib-dot-dll** [comment]
The log https://flathub.org/builds/api/v2/logs/84615/raw said that
```
  PNG               --with-png=yes		yes
```
:thinking:

**Comment by nedrichards** [comment]
bot, build com.visualstudio.code-oss

**Comment by amtlib-dot-dll** [comment]
No, it also fails. The last resort would be convert the images on my PC and then upload it

If your builder supports interruption you may just stop this build :joy:

**Comment by nedrichards** [comment]
fwiw converting and offering icons manually is something I've done before in other apps.

**Comment by amtlib-dot-dll** [comment]
The PNG issue should be solved. I am now testing on my machine

**Comment by nedrichards** [comment]
bot, build com.visualstudio.code-oss

**Comment by amtlib-dot-dll** [comment]
This version https://github.com/flathub/flathub/pull/272/commits/08e32bb68fe2ebdb2133aaf341d8ff2784ad03c2 builds and runs OK on my machine

**Comment by amtlib-dot-dll** [comment]
Good morning, is there anything else to do?

**Comment by amtlib-dot-dll** [comment]
I think there is a bug in `flatpak-builder` that the cache dir should be clean or the build would fail.

**Comment by nedrichards** [comment]
Just thought, you should add the electron base app since that's what it's
there for rather than just gconf etc.

**Comment by nedrichards** [comment]
bot, build com.visualstudio.code-oss

**Comment by amtlib-dot-dll** [comment]
The last build you triggered is already based on the electron base app, thank you for your advice :smile:

File: flathub.json (hunk 4 above)
Comment (amtlib-dot-dll) [reply flathub.json]: 32-bit support added

File: com.visualstudio.code-oss.appdata.xml [hunk 5]
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<component type="desktop">
```
Comment (TingPing) [inline com.visualstudio.code-oss.appdata.xml]: Needs `<releases>`

File: com.visualstudio.code-oss.json [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+    "sdk": "org.freedesktop.Sdk",
+    "command": "code-oss",
+    "separate-locales": false,
+    "finish-args": [
+        "--share=ipc",
+        "--socket=x11",
+        "--socket=pulseaudio",
+        "--share=network",
+        "--device=dri",
+        "--filesystem=home",
```
Comment (TingPing) [inline com.visualstudio.code-oss.json]: Probably want `host`

**Comment by amtlib-dot-dll** [comment]
Modified accordingly to @TingPing's comment

**Comment by amtlib-dot-dll** [comment]
There seems to be a bug in the runtime shipped Python as described [here](https://stackoverflow.com/q/28890754). In the build process a Python script (`gyp_main.py`) were invoked multiple times, but some invocations failed with `EOFError: EOF read where object expected` when importing the `gyp.input` module, while other invocations succeeded. These modules are expected to be read-only, and this inconsistency makes me puzzled. [It is fixed in Python 3.3](https://stackoverflow.com/a/37083837) but `gyp` requires Python 2 and I don't know is there a patched version of Python 2 without such an issue.

**Comment by amtlib-dot-dll** [comment]
I have to disable the 32-bit build. The build system of visual studio code complains that "JavaScript heap out of memory" when run under 32-bit. The 32-bit binary could probably only be cross-built under 64-bit builders.

**Comment by nedrichards** [comment]
bot, build com.visualstudio.code-oss

**Comment by amtlib-dot-dll** [comment]
Sorry, the patch won't fix the `ripgrep` based search. One workaround is to disable `ripgrep` and use JavaScript-based search in the settings. I am still digging into the problem.

**Comment by nedrichards** [comment]
OK, cool. I haven't looked into it but I assume that the binary dependencies don't build arm or aarch64 versions?

**Comment by nedrichards** [comment]
bot, build com.visualstudio.code-oss

**Comment by amtlib-dot-dll** [comment]
The reason is the wrong version of the downloaded `rg` binary. Fixed. Please rebuild thanks :smiley:

**Comment by nedrichards** [comment]
bot, build com.visualstudio.code-oss

**Comment by amtlib-dot-dll** [comment]
The ID have been renamed to `com.visualstudio.code.oss` according to upstream

**Comment by nedrichards** [comment]
bot, build com.visualstudio.code.oss

**Comment by amtlib-dot-dll** [comment]
Sorry, bug found, please rebuild :joy:

**Comment by nedrichards** [comment]
bot, build com.visualstudio.code.oss

File: com.visualstudio.code.oss.json [hunk 7]
```diff
@@ -16,7 +16,8 @@
         "--device=dri",
         "--filesystem=host",
         "--persist=.vscode-oss",
-        "--talk-name=org.freedesktop.Notifications"
+        "--talk-name=org.freedesktop.Notifications",
+        "--env=GTK_IM_MODULE=xim"
```
Comment (nedrichards) [inline com.visualstudio.code.oss.json]: @TingPing I'm not an input method expert, will this break the built in ibus support?

Comment (amtlib-dot-dll) [reply com.visualstudio.code.oss.json]: I don't know whether it breaks, but I just discovered that the IME won't work with our without this environment variable :joy: :joy: @TingPing What should we do to solve this problem? 🤔 https://github.com/flathub/com.visualstudio.code/issues/24 and https://github.com/flathub/io.atom.Atom/issues/6

Comment (nedrichards) [reply com.visualstudio.code.oss.json]: ibus seems to be the only supported IME within flatpak currently. But just to clarify, are you saying that ibus doesn't work either? As mentioned in https://github.com/flathub/org.signal.Signal/issues/17 could you try building and using [the ibus testcase flatpak](https://github.com/flathub/org.signal.Signal/issues/17#issuecomment-368955752)?

Comment (amtlib-dot-dll) [reply com.visualstudio.code.oss.json]: IME works in the testcase flatpak. What should I add to my recipe? :thinking:

Comment (nedrichards) [reply com.visualstudio.code.oss.json]: Not sure currently - but this very much seems to be a common issue for all electron apps.

Comment (amtlib-dot-dll) [reply com.visualstudio.code.oss.json]: :woman_facepalming:

**Comment by amtlib-dot-dll** [comment]
@nedrichards Excuse me but is there anything other to do to get this recipe included in flathub? Thanks

File: build.sh [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    sed -i "s/'vscode\-api\-tests',//" build/gulpfile.vscode.js
+    sed -i "s/'vscode\-colorize\-tests',//" build/gulpfile.vscode.js
+    sed -i "s/'vscode\-api\-tests',//" build/npm/postinstall.js
+    sed -i "s/'vscode\-colorize\-tests',//" build/npm/postinstall.js
+    python2 /app/lib/node_modules/npm/node_modules/node-gyp/gyp/gyp_main.py --help || true
+    npm_config_tarball="$(realpath ../misc/iojs-v1.7.9.tar.gz)" yarn install --offline --verbose --frozen-lockfile
+    rm node_modules/vscode-ripgrep
+    cp -r ../vscode-ripgrep-0.7.1-patch.0.1 node_modules/vscode-ripgrep
+    echo "/// <reference types='@types/node'/>" > extensions/emmet/src/typings/refs.d.ts
+    node_modules/.bin/gulp vscode-linux-$(node -e 'console.log(process.arch)')-min $([ $FLATPAK_ARCH == 'x86_64' ] && echo '--max_old_space_size=4096')
```
Comment (nedrichards) [inline build.sh]: > $FLATPAK_ARCH == 'x86_64' 

This won't help for other architectures.

File: com.visualstudio.code.oss.json [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+    "finish-args": [
+        "--share=ipc",
+        "--socket=x11",
+        "--socket=pulseaudio",
+        "--share=network",
+        "--device=dri",
+        "--filesystem=host",
+        "--persist=.vscode-oss",
+        "--talk-name=org.freedesktop.Notifications",
+        "--env=GTK_IM_MODULE=xim"
```
Comment (nedrichards) [inline com.visualstudio.code.oss.json]: You need to remove this variable.

File: build.sh (hunk 8 above)
Comment (amtlib-dot-dll) [reply build.sh]: Yes. The build system won't work under `i386` for limited memory.

File: com.visualstudio.code.oss.json (hunk 9 above)
Comment (amtlib-dot-dll) [reply com.visualstudio.code.oss.json]: Removed. Thank you for your advice

**Comment by amtlib-dot-dll** [comment]
I'm not sure whether `aarch64` works, and I don't know how to test it, but I know that `arm` probably won't work for the same reason as `i386`, that the build system needs more than 4 GiB of memory.

**Comment by cheese** [comment]
Why not make this as a branch of the current vscode flatpak?

**Comment by amtlib-dot-dll** [comment]
@cheese 
The [build system](https://flathub.org/builds) does not support non-master branch

**Comment by nedrichards** [comment]
@amtlib-dot-dll well, the aarch64 builder is actually the most powerful server flathub has right now... I believe it has 128Gb of RAM, so that won't be a problem.

**Comment by nedrichards** [comment]
bot, build com.visualstudio.code.oss

**Comment by amtlib-dot-dll** [comment]
@nedrichards By the way, does the build system support cross-compiling? Therefore we can make real use of the most powerful server and we will have 32-bit builds. As we know few personal workstations run on `aarch64`

**Comment by amtlib-dot-dll** [comment]
`arm` is limited to a memory less than 4 GiB for the same reason as `i386` despite the real RAM size.

**Comment by nedrichards** [comment]
Yeah - that's a good point (on arm) I know you can compile stuff in other architectures using flatpak I haven't really looked into it with more detail.

**Comment by amtlib-dot-dll** [comment]
No. Currently `flatpak-builder` doesn't support the cross compiling needed by this recipe

What `flatpak-builder` does when building for arch T on an machine of arch S:
1. Create an environment of T. If S is `x86_64` and T is `i386`, then it's easy; if S is `aarch64` and T is `i386`, `qemu` will be needed
2. Run the toolchain in the environment T

This recipe won't work when T is `i386`, because of the memory limit

The current build infrastructure is far from cross compiling. What I want is a toolchain to *run* the compiler in S and the compiler just *generates* binaries for T without running them. For S ∈ {`x86_64`, `i386`, `aarch64`, `arm`} and T ∈ {`x86_64`, `i386`, `aarch64`, `arm`}, 4 × 4 = 16 types of tool chains will be needed

**Comment by nedrichards** [comment]
Flathub supports aarch64 and we'd like all software that we can to be available on every architecture.. Saying that it's not terribly popular isn't a compelling argument, we start from enabling everything and disable what can't work.

**Comment by amtlib-dot-dll** [comment]
Another problem that prohibits `aarch64` builds is that the [upstream](https://github.com/Microsoft/vscode/blob/1.20.1/build/gulpfile.vscode.js#L379-L381) does not provide it

**Comment by amtlib-dot-dll** [comment]
New version 1.21.1

**Comment by nedrichards** [comment]
bot, build com.visualstudio.code.oss

**Comment by nedrichards** [comment]
thanks - let's have a look - and then prepare to revert and merge as and when it fails! ;-)

**Comment by amtlib-dot-dll** [comment]
The 32-bit build fails with the same symptom in my worker, segmentation fault.

**Comment by nedrichards** [comment]
yeah, back to x86-64 only then!

**Comment by amtlib-dot-dll** [comment]
By the way, how long will it take to be included in flathub? 🤔 Thanks

**Comment by nedrichards** [comment]
Now we've tested everything and done the revert I'll rebuild again and push
it on Monday.

**Comment by nedrichards** [comment]
bot, build com.visualstudio.code.oss

**Comment by nedrichards** [comment]
Repository has been created: https://github.com/flathub/com.visualstudio.code.oss

**Pull request closed without merging**